[WIP] ipq40xx: restructure mikrotik 5acd and add lhg-5-ac

PolynomialDivision opened the pull request:
Needs testing of
- [ ] MikroTik SXTsq 5 AC
- [ ] MikroTik LHG 5 ac
- [ ] MikroTik hAP ac2

This is a rework of: https://github.com/openwrt/openwrt/pull/4266

PolynomialDivision committed:
ipq40xx: add dtsi to reduce code duplication

Add "qcom-ipq4018-mikrotik-5acd.dtsi" and rework:
- sxtsq-5-ac
- hap-ac2

Both devices are very similar.

This PR is based on the work of William Wortel <wwortel@dorpstraat.com>.

Signed-off-by: Nick Hainke <vincent@systemli.org>

PolynomialDivision committed:
ipq40xx: add MikroTik LHG 5 ac (RBLHGG-5acD)

This adds support for the MikroTik LHG 5 ac (RBLHGG-5acD). It is an
outdoor device with an integrated dual polarization 24.5 dBi grid
antenna for long range connection.

See https://mikrotik.com/product/lhg_5_ac for more info.

Specifications:
 - SoC: Qualcomm Atheros IPQ4018
 - RAM: 256 MB
 - Storage: 16 MB NOR
 - Wireless: 802.11a/n/ac, 24.5 dBi antenna, 2 chains
 - Ethernet: 10/100,1000, passive PoE

Installation:
Boot the initramfs image via TFTP and then flash the sysupgrade
image using "sysupgrade -n".
See: https://openwrt.org/toh/mikrotik/common

This PR is based on the work of William Wortel <wwortel@dorpstraat.com>.

Signed-off-by: Nick Hainke <vincent@systemli.org>

chunkeey commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4018-mikrotik-5acd.dtsi:
```diff
@@ -0,0 +1,236 @@
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+	model = "MikroTik 5acD";
```
I think you could do without the ````model````and ````compatible```` in the .dtsi root node there.
(They will get overwritten anyway).

PolynomialDivision replied:
thanks! removed it.

PolynomialDivision committed:
ipq40xx: move product codes to DEVICE_ALT0_MODEL

The ipq40xx MikroTik targets have their product code in their names.
Add "RBD52G-5HacD2HnD-TC" to the hAP ac².

Move product codes from DEVICE_MODEL to DEVICE_ALT0_MODEL.

Signed-off-by: Nick Hainke <vincent@systemli.org>

adschm commented:
I'd prefer not to have special characters like the superscript "2" in commit title/message.

adschm commented:
> ipq40xx: add dtsi to reduce code duplication

This should be more specific: Which devices are you dealing with? The reason "code duplication" is not so important here and could be moved into the message part...

adschm commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -44,6 +44,7 @@ ALLWIFIBOARDS:= \
 	linksys_mr8300-v0 \
 	luma_wrtq-329acn \
 	mikrotik_hap-ac2 \
+  mikrotik_lhg-5-ac \
```
Broken indent.

adschm commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -130,6 +131,7 @@ $(eval $(call generate-ipq-wifi-package,linksys_mr8300-v0,Linksys MR8300))
 $(eval $(call generate-ipq-wifi-package,luma_wrtq-329acn,Luma WRTQ-329ACN))
 $(eval $(call generate-ipq-wifi-package,mikrotik_hap-ac2,Mikrotik hAP ac2))
 $(eval $(call generate-ipq-wifi-package,mikrotik_sxtsq-5-ac,MikroTik SXTsq 5 ac))
+$(eval $(call generate-ipq-wifi-package,mikrotik_lhg-5-ac,MikroTik LHG 5 ac))
```
Sorting.

adschm commented on target/linux/ipq40xx/image/mikrotik.mk:
```diff
@@ -19,6 +19,14 @@ define Device/mikrotik_hap-ac2
 
+define Device/mikrotik_lhg-5-ac
+	$(call Device/mikrotik_nor)
+	DEVICE_MODEL := LHG 5 ac (RBLHGG-5acD)
```
Please remove the brackets and use DEVICE_ALT0_VENDOR and *_MODEL instead.

PolynomialDivision replied:
This is just the product code for the device and not an alternative model? I can also just remove it completely?

adschm replied:
This here is exactly what the DEVICE_ALT* syntax was developed for. It's meant for cases where a single device has two different names, i.e. to get rid of the brackets in DEVICE_MODEL.

Everything else like using the same recipe for different devices was actually an abuse of this initial concept.

rogerpueyo replied:
Hi,

As I understood it (please correct me if I'm wrong), the ALT* syntax was used for identical devices with different manufacturer and/or name, like:
```
  DEVICE_VENDOR := AirTight Networks
  DEVICE_MODEL := C-75
  DEVICE_ALT0_VENDOR := Mojo Networks
  DEVICE_ALT0_MODEL := C-75
  DEVICE_ALT1_VENDOR := WatchGuard
  DEVICE_ALT1_MODEL := AP320
```
or different versions with identical hardware:
```
  DEVICE_MODEL := TL-WR941ND
  DEVICE_VARIANT := v2/v3
  DEVICE_ALT0_VENDOR := TP-Link
  DEVICE_ALT0_MODEL := TL-WR941N
  DEVICE_ALT0_VARIANT := v2/v3
```

I am not sure if this is the case here, with MikroTik devices. This brand [most often] uses a short "commercial" name for their devices and also assigns them a "product code". For instance, "SXTsq 5 ac" (commercial) and "RBSXTsqG-5acD" (product code) (see https://mikrotik.com/product/sxtsq_5_ac). Sometimes, however, the same commercial name is used for devices with equivalent features but totally different hardware (e.g., ath79 "wAP AC (RBwAPG-5Hac**T**2HnD)" and ipq40xx "wAP AC RBwAPG-5Hac**D**2HnD"

In my opinion, it is good to use both commercial and product code here, and also in the DTS file, so that e.g., /proc/cpuinfo displays it.

Just my 2 cents.

adschm commented on target/linux/ipq40xx/image/mikrotik.mk:
```diff
@@ -12,7 +12,7 @@ endef
 define Device/mikrotik_hap-ac2
 	$(call Device/mikrotik_nor)
-	DEVICE_MODEL := hAP ac2
+	DEVICE_MODEL := hAP ac2 (RBD52G-5HacD2HnD-TC)
```
DEVICE_ALT0_* ...

rogerpueyo reviewed (commented):
/OpenWrt/qca-swiss-army-knife/tools/scripts/ath10k/ath10k-bdencoder -i board-mikrotik_lhg-5-ac.qca4019 
```
FileSize: 12176
FileCRC32: 0ad8a73e
FileMD5: dbc68c769ab9e54167f5772e6757f5fe
BoardNames[0]: 'bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=MikroTik-7442a-LHG5ACD'
BoardLength[0]: 12064
BoardCRC32[0]: 3514aeae
BoardMD5[0]: d258c7dda3109b29344d15f76a7ca98b
```
Is the ath10k driver correctly loading the file? If I'm not wrong, given line
```
 $(eval $(call generate-ipq-wifi-package,mikrotik_lhg-5-ac,MikroTik LHG 5 ac))
```
it should be expecting variant to be "MikroTik-7442a-LHG5ACD".

john-tho commented:
When you do get your device to test, could you please try the sxtsq5 ac image on it? I expect it will work.
The OEM uses the same FCC IDs for these devices: SXTsq 5 ac, LDF, LHG 5 ac, LHG XL 5 ac.
I can see that my sxtsq5 ac uses `model = "LHG 5ac";` in the OEM DTB.

The model information is available from the OEM hard config data:
```
root@link:~# cat /sys/firmware/mikrotik/hard_config/product_name 
SXTsq 5 ac
root@link:~# cat /sys/firmware/mikrotik/hard_config/board_product_code 
RBSXTsqG-5acD
root@link:~# cat /sys/firmware/mikrotik/hard_config/board_identifier 
lhg-dk
```
Is it worth building 4 images with only the model name different?

PolynomialDivision commented:
There are a lot of ipq40xx PRs. I will close that and reopen again later after all the other more important changes are applied.
